### PR TITLE
Pass `librdkafka` debug level down as a value

### PR DESCRIPTION
**Pull request by guswynn**
This pr implements the changes described in https://github.com/MaterializeInc/materialize/issues/10441 (minus one). 
It accomplishes this by:
- adding more helper functions
- pre-determining the `tracing::Level` for `librdkafka` based on the `Targets` for the fmt layer
- passing this level down. We have a few `Context` structs, mainly `CoordContext` so that this level can be passed to purification, and `RenderContext` to be passed into source and sink creation
  - In a later pr, more information (likely the new aws information adde by I think @quodlibetor ) can be moved into these contexts, I believe
- Passing this level into the new `ClientConfig::new` wrapper, and `conditionally enabling `debug=all`. It is not 100% to me what the difference between these is, but I believe we should leave it as is for now and investigate later, to prevent behavioral changes
What is not accomplished:
- removing the filter from the metrics layer. In my testing, the regression existed in at least the `KafkaUpsert` feature-benchmark. This benchmark is pretty noisy, but after some investigation, I believe the regression is due to us not longer hitting this fast path: https://docs.rs/tracing/latest/src/tracing/macros.rs.html#2064-2069 that is checked before every tracing call.
  - Its possible that turning off `tracing-log` will help, but I doubt removing this filter will ever be perf-neutral: We are simply 1. checking `enabled` more and 2. dispatching more events (i should have realized this would have happened)
  - Even with the regression, this change is worth it, as it still means the `debug=all` case is correctly setup
  - I think in the platform world, I think we will have to accept this regression, especially because turning it on `tokio-console`, and, more importantly, `opentelemetry` will cause it to exist anyways. We probably need to have some kind of proposal on the default, single-instance setup for mz, vs, the cloud/platform setup, which configures tracing in a more aggressive way.
  - @benesch @philip-stoev it would be nice to get some pointers, after this lands, on how to do a full perf test when removing the filter from other other layers
  
Overall, I think this pr improves the general flow of values and extra context into sources, and fixes the mis-use of `tracing::enabled!` for when we DO have more layers enabled (which may only be the `tokio-console` option right now), so i think it is correct to land it as is

### Motivation

Which of the following best describes the motivation behind this PR?

   * This PR refactors existing code.
as described above and in https://github.com/MaterializeInc/materialize/issues/10441

### Tips for reviewer

- bottom pr: refactor so that in the later commit, `ClientConfig::new` is always called correctly (through ore)
- second pr: the main refactor, described above:
- third pr: this is a description of an absolutely bizarre bug, and after multiple hours with the tracing team, the conclusion was: lets not use `tracing-log` anymore (we are VERY close to this). I think `rdkafka` is the last MEANINGFUL dep that used `log`
- top pr: a TODO describing the performance followup

### Testing

- [N/A] This PR has adequate test coverage / QA involvement has been duly considered.
tested locally

### CODE REVIEW

**Review by benesch — COMMENTED** [review]
Thanks for pushing on this! I think we need @frankmcsherry or @petrosagg to weigh in on whether `RenderContext` still makes sense, or whether the librdkafka level ought just get attached to `StorageState` and `ComputeState`.

**Review by frankmcsherry — CHANGES_REQUESTED** [review]
Design-wise, it looks like `librdkafka_debug` wants to live at least as far down as `StorageState`, not in `Worker`. I'd personally it to live in source and sink connectors instead, as workers/compute and potentially storage are otherwise Kafka-agnostic.

If there is a reason to have it at the root of `Worker` lmk.

**Comment by guswynn** [comment]
@frankmcsherry as far as I can tell, it needs to be passed down into the `ActiveComputeState` to be passed to kafka sinks, which is created by the `Worker` in `activate_compute`

**Comment by guswynn** [comment]
I can see it being better to move it into structs that are for extra context for compute and storage

**Comment by benesch** [comment]
Did you mean to abandon this, Gus? I think Frank's proposal was just to plumb the bool via `KafkaConnector` rather than `StorageState`.

**Comment by benesch** [comment]
> I don't think connectors are the right place for this: they are serde::Serializable, stable representations of sources, whereas librdkafka_debug is a property of a running process. It is similar to aws_external_id, which is moved directly from mz_dataflow::server::Config into StorageState. I can do the same for librdkafka_debug there, but for ComputeState, which is generated after the Worker is created, the Worker object itself will need to hold onto this "extra context". I factored that into SharedComputeState to clean up how it looks!

This is a very good point. This wasn't the case until recently, but it *certainly* is now. I buy this argument.

File: src/coord/src/coord.rs [hunk 1]
```diff
@@ -257,6 +257,14 @@ pub struct Config {
     pub metrics_registry: MetricsRegistry,
     pub persister: PersisterWithConfig,
     pub now: NowFn,
+    pub librdkafka_debug: tracing::Level,
+}
+
+/// Extra context to pass through to connector creation
```
Comment (benesch) [inline src/coord/src/coord.rs:263]: Stale now?

File: src/compute/src/compute_state.rs [hunk 2]
```diff
@@ -38,6 +38,15 @@ use crate::logging;
...
 use crate::sink::SinkBaseMetrics;
 
+/// Extra context shared within a single `Worker`
+/// instance.
+/// Should be kept cheaply-cloneable.
+#[derive(Clone)]
+pub struct SharedComputeState {
+    /// The debug level to set librdkafka to
+    pub librdkafka_debug: tracing::Level,
+}
```
Comment (benesch) [inline src/compute/src/compute_state.rs:48]: nit: "shared" makes me think `Arc`, but this is just directly cloned.

**Review by benesch — COMMENTED** [review]
I took a look, and this all checks out for me. Deferring to @frankmcsherry, who might have some feedback about whether each one of these `Context` structs is worth extracting, or whether there are places where the `librdkafka_debug` field should just be embedded directly in an existing struct.

File: src/coord/src/coord.rs (hunk 1 above)
Comment (guswynn) [reply src/coord/src/coord.rs:263]: This is used in `coord::sink_connector`?

File: src/compute/src/compute_state.rs (hunk 2 above)
Comment (guswynn) [reply src/compute/src/compute_state.rs:48]: "cheaply-cloneable" to me means it contains `Copy` types or `Arc`-wrapped types

Comment (guswynn) [reply src/compute/src/compute_state.rs:48]: is `CommonComputeState` reasonable?

File: src/compute/src/sink/avro_ocf.rs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                    Ok(())
+                };
+
+                if let Err(e) = fallible() {
+                    error!("{}", e);
+                }
+            })
+        },
+    )
+}
```
Comment (benesch) [inline src/compute/src/sink/avro_ocf.rs:135]: This seems to have been resurrected from the dead! Rebase glitch?

File: src/ore/src/process.rs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+/// Returns a human-readable version of how the current process was invoked.
+pub fn invocation() -> String {
+    env::vars_os()
+        .map(|(name, value)| {
+            (
+                name.to_string_lossy().into_owned(),
+                value.to_string_lossy().into_owned(),
+            )
+        })
+        .filter(|(name, _value)| name.starts_with("MZ_"))
```
Comment (benesch) [inline src/ore/src/process.rs:26]: The `MZ_` bit here ought not be hardcoded!

File: src/ore/src/tracing.rs [hunk 5]
```diff
@@ -141,11 +141,11 @@ pub struct TracingConfig<'a> {
 
 /// Configures tracing according to the provided command-line arguments.
 /// Returns a `Write` stream that represents the main place `tracing` will
-/// log to.
+/// log to, and a `tracing::Level` that is the `librdkafka` debug level.
```
Comment (benesch) [inline src/ore/src/tracing.rs:144]: To keep abstraction boundaries clean, `ore` should not know anything about `librdkafka`. Is there something else we can return here instead? The filter maybe?

Comment (guswynn) [reply src/ore/src/tracing.rs:144]: We could take in a list of `targets` and return a list of `Level`s for those targets we checked?

File: src/ore/src/process.rs (hunk 4 above)
Comment (guswynn) [reply src/ore/src/process.rs:26]: ah yes, ore, i can have it as an optional parameter

**Comment by guswynn** [comment]
I believe this pr is ready to review again! It only passes the value into the coordinator to do purification, and storaged, computed we default to `INFO` as sinks are moving into storage eventually

**Comment by benesch** [comment]
I promise I'll handle all the rebase conflicts that I'm introducing here! #12402 is the latest in a series of PRs to simplify our tracing setup by removing optionality that won't matter in Materialize Platform.

**Comment by guswynn** [comment]
@benesch thanks!

**Comment by guswynn** [comment]
closed by #12505

**Pull request closed without merging**